File: Python/bucketsort.py

def bucketSort(array):
    bucket = []

    for i in range(len(array)):
        bucket.append([])

    for j in array:
        index_b = int(10 * j)
        bucket[index_b].append(j)

    for i in range(len(array)):
        bucket[i] = sorted(bucket[i])

    k = 0
    for i in range(len(array)):
        for j in range(len(bucket[i])):
            array[k] = bucket[i][j]
            k += 1
    return array


print("Sorted Array in descending order is")
print(bucketSort(array))



"""
Problem : Sort the given array using bucket sort.

Sort the elements by first dividing the elements into several groups or buckets. The elements inside each bucket are sorted using any of the suitable sorting algorithms or recursively calling the same algorithm.

Sample I/O:

INPUT:
3
4
5
2
1
Sorted Array in descending order is
[1, 2, 3, 4, 5]


Time complexity: O(n^2) (worst)
Space complexity: Ω(n+k) (best)
"""

File: C-Plus-Plus/Bucketsort.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

void bucketSort(vector<float> &arr, int n)
{
    vector<float> b[n];

    for (int i=0; i<n; i++)
    {
       int x = n*arr[i];
       b[x].push_back(arr[i]);
    }

    for (int i=0; i<n; i++)
       sort(b[i].begin(), b[i].end());

    int index = 0;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < b[i].size(); j++)
          arr[index++] = b[i][j];
}

int main()
{
    int n;
    cout<<"Enter no of Elements: ";
    cin>>n;
    cout<<"\nEnter Elements";
    vector<float> arr(n);
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    cout<<"\nBefore Sorting\n";
    for (int i=0; i<n; i++)
    cout << arr[i] << " ";

    bucketSort(arr, n);

    cout << "\nAfter Sorting \n";
    for (int i=0; i<n; i++)
       cout << arr[i] << " ";
    return 0;
}

/*
Worst Case Complexity: O(n2)
Best Case Complexity: O(n+k)
Average Case Complexity: O(n)
Space Complexity: O(n+k)
Input and Output:
Enter no of Elements: 8
Enter Elements
0.235
0.101
0.476
0.7645
0.15
0.142
0.7845
0.4545
Before Sorting:
0.235 0.101 0.476 0.7645 0.15 0.142 0.7845 0.4545
After Sorting
0.101 0.142 0.15 0.235 0.4545 0.476 0.7645 0.7845
*/
